File: C++/2684. Maximum Number of Moves in a Grid.cpp
class Solution {
public:
    int rec(int i,int j,vector<vector<int>>& grid,vector<vector<int>>& dp){
        if(dp[i][j]!=-1) return dp[i][j];
        int ans = 0;
        for(int k = -1;k<=1;k++){
            int a = i+k;
            int b = j+1;
            if(a>=grid.size() || a<0 || b>=grid[0].size()) continue;
            if(grid[i][j]<grid[a][b]){
                ans = max(ans,1+rec(a,b,grid,dp));
            }
        }
        return dp[i][j]=ans;
    }
    int maxMoves(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        int ans = 0;
        vector<vector<int>> dp(n+1,vector<int>(m+1,-1));
        for(int i=0;i<n;i++){
           ans = max(ans,rec(i,0,grid,dp));
        }
        return ans;
    }
};

File: JAVA/2684. Maximum Number of Moves in a Grid.java
class Solution {
    
    public int maxMoves(int[][] grid) {
     int m = grid.length;
     int n = grid[0].length;
     int[][] dp = new int[m][n];
     int ans = 0;
     for(int j=n-2;j>=0;j--)
     {
        for(int i=0;i<m;i++)
        {
            if(i>0 && grid[i][j]<grid[i-1][j+1])
            dp[i][j] = Math.max(dp[i][j],1+dp[i-1][j+1]);

            if(grid[i][j]<grid[i][j+1])
            dp[i][j] = Math.max(dp[i][j],1+dp[i][j+1]);

            if(i!=m-1 && grid[i][j]<grid[i+1][j+1])
            dp[i][j] = Math.max(dp[i][j],1+dp[i+1][j+1]);

            if(j==0)
            ans = Math.max(ans,dp[i][j]);
        }
     }
     return ans;
    }
}
